File: sorting/counting_sort/counting_sort.cpp
#include <algorithm>
#include <vector>

void counting_sort(std::vector<int>& vec, int min_value, int max_value) {
  auto temp = std::vector(vec);
  auto cnt = new int[max_value - min_value + 1]();

  // 카운팅 작업
  for (int element : vec) {
    cnt[element - min_value]++;
  }

  // 누적 합 작업
  for (size_t i = 1; i <= max_value - min_value; i++) {
    cnt[i] += cnt[i - 1];
  }

  // 카운트 배열 참조해 알맞은 위치 찾아서 넣기
  for (size_t i = temp.size(); i > 0; i--) {
    size_t idx = temp[i - 1] - min_value;
    vec[--cnt[idx]] = temp[i - 1];
  }

  delete[] cnt;
}

File: sorting/counting_sort/counting_sort.rs
fn counting_sort(array: &mut [i32], min_value: i32, max_value: i32) {
    let temp = array.to_vec();
    let mut cnt = vec![0usize; (max_value - min_value + 1) as usize];

    // 카운팅 작업
    for element in array.iter() {
        cnt[(*element - min_value) as usize] += 1;
    }

    // 누적 합 작업
    for i in 1..cnt.len() {
        cnt[i] += cnt[i - 1];
    }

    // 카운트 배열 참조해 알맞은 위치 찾아서 넣기
    for element in temp.iter().rev() {
        let idx = (*element - min_value) as usize;
        cnt[idx] -= 1;
        array[cnt[idx]] = *element;
    }
}